Preliminary support for OpenBSD in the stdlib.

3405691582 opened the pull request:
Preliminary support for OpenBSD in the stdlib.
    
These should hopefully all be uncontroversial, minimal changes to deal
with progressing the build to completion on OpenBSD or addressing minor
portability issues. This is not the full set of changes to get a
successful build; other portability issues will be addressed in future
commits.
    
Most of this is just adding the relevant clauses to the ifdefs, but of
note in this commit:
    
* StdlibUnittest.swift: the default conditional in _getOSVersion assumes
  an Apple platform, therefore the explicit conditional and the relevant
  enums need filling out. The default conditional should be #error, but
  we'll fix this in a different commit.
    
* tgmath.swift.gyb: inexplicably, OpenBSD is missing just lgammal_r.
  Tests are updated correspondingly.
    
* ThreadLocalStorage.h: we use the pthread implementation, so it
  seems we should typedef __swift_thread_key_t as pthread_key_t.
  However, that's also a tweak for another commit.
    
* libc-openbsd.modulemap.gyb: separate libc modulemap for OpenBSD.
  sdk in CMake will be set in a future commit, but it should be safe to add
  this here now.

3405691582 committed:
Preliminary support for OpenBSD in the stdlib.

These should hopefully all be uncontroversial, minimal changes to deal
with progressing the build to completion on OpenBSD or addressing minor
portability issues. This is not the full set of changes to get a
successful build; other portability issues will be addressed in future
commits.

Most of this is just adding the relevant clauses to the ifdefs, but of
note in this commit:

* StdlibUnittest.swift: the default conditional in _getOSVersion assumes
  an Apple platform, therefore the explicit conditional and the relevant
  enums need filling out. The default conditional should be #error, but
  we'll fix this in a different commit.

* tgmath.swift.gyb: inexplicably, OpenBSD is missing just lgammal_r.
  Tests are updated correspondingly.

* ThreadLocalStorage.h: we use the pthread implementation, so it
  seems we should typedef __swift_thread_key_t as pthread_key_t.
  However, that's also a tweak for another commit.

compnerd commented on stdlib/public/Platform/glibc.modulemap.gyb:
```diff
@@ -19,14 +19,14 @@
 /// future official glibc modulemap.
 module SwiftGlibc [system] {
-% if CMAKE_SDK in ["LINUX", "FREEBSD", "CYGWIN"]:
+% if CMAKE_SDK in ["LINUX", "FREEBSD", "OPENBSD", "CYGWIN"]:
```
Might be nicer to split out the file separately for FreeBSD's libc.

3405691582 replied:
To clarify: you mean like how there's a separate modulemap for bionic, but still called SwiftGlibc?

compnerd replied:
Yeah, exactly so.

3405691582 replied:
Sounds good. I'll pull the modulemap changes into a separate commit (in this PR).

compnerd commented:
@swift-ci please test

3405691582 committed:
[stdlib] Modulemap for libc on OpenBSD.

Provide a separate modulemap for OpenBSD's libc, rather than try to
continue to overload the modulemap for glibc, since glibc has its own
idiosyncracies.

broadwaylamb commented on stdlib/private/SwiftPrivateThreadExtras/SwiftPrivateThreadExtras.swift:
```diff
@@ -17,7 +17,7 @@
 #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
 import Darwin
-#elseif os(Linux) || os(FreeBSD) || os(PS4) || os(Android) || os(Cygwin) || os(Haiku)
+#elseif os(Linux) || os(FreeBSD) || os(OpenBSD) || os(Android) || os(Cygwin) || os(Haiku)
```
You've accidentally replaced `os(PS4)` with `os(OpenBSD)` instead of adding the latter.

```suggestion
#elseif os(Linux) || os(FreeBSD) || os(OpenBSD) || os(PS4) || os(Android) || os(Cygwin) || os(Haiku)
```

3405691582 replied:
Sorry about that. Accepted fix.

broadwaylamb replied:
No worries! Thank you for your effort.

3405691582 committed:
Fix typo in SwiftPrivateThreadExtras.

Co-Authored-By: Sergej Jaskiewicz <broadwaylamb@users.noreply.github.com>

xwu commented on stdlib/private/SwiftPrivateThreadExtras/ThreadBarriers.swift:
```diff
@@ -12,7 +12,7 @@
 #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
 import Darwin
-#elseif os(Linux) || os(FreeBSD) || os(PS4) || os(Android) || os(Cygwin) || os(Haiku)
+#elseif os(Linux) || os(FreeBSD) || os(PS4) || os(Android) || os(Cygwin) || os(Haiku) || os(OpenBSD)
```
```suggestion
#elseif os(Linux) || os(FreeBSD) || os(OpenBSD) || os(PS4) || os(Android) || os(Cygwin) || os(Haiku)
```

...maybe, just for consistency with the rest of these? (Inconsequential nit.)

3405691582 replied:
Good catch.

3405691582 committed:
Ensure consistency of ordering in preprocessor conditionals.

Co-Authored-By: Xiaodi Wu <13952+xwu@users.noreply.github.com>

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test Linux platform

compnerd commented:
LGTM, seems all the review comments have been addressed.  Lets get OpenBSD moving along :)
